File: test/tunnel_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:httpku/services/tunnel_engine.dart';
import 'package:httpku/models/tunnel_config.dart';

void main() {
  group('TunnelEngine Tests', () {
    test('Initial status is disconnected and logs empty', () {
      final engine = TunnelEngine();
      expect(engine.status, TunnelStatus.disconnected);
      expect(engine.isConnected, isFalse);
      expect(engine.logs, isEmpty);
    });

    test('Start and stop tunnel changes state and records logs', () async {
      final engine = TunnelEngine();
      expect(engine.status, TunnelStatus.disconnected);

      final startFuture = engine.startTunnel();
      expect(engine.isConnecting || engine.isConnected, isTrue);

      await startFuture;
      expect(engine.status, TunnelStatus.connected);
      expect(engine.isConnected, isTrue);
      expect(engine.logs, isNotEmpty);

      await engine.stopTunnel();
      expect(engine.status, TunnelStatus.disconnected);
      expect(engine.isConnected, isFalse);
      expect(engine.logs.last.message, contains('disconnected'));
    });

    test('Clear logs empties log list', () {
      final engine = TunnelEngine();
      engine.addLog('Test log entry');
      expect(engine.logs.length, 1);

      engine.clearLogs();
      expect(engine.logs, isEmpty);
    });

    test('Config updates correctly', () {
      final engine = TunnelEngine();
      final newConfig = TunnelConfig(remoteAddr: 'ssh.server.com', remotePort: 22);
      engine.updateConfig(newConfig);

      expect(engine.config.remoteAddr, 'ssh.server.com');
      expect(engine.config.remotePort, 22);
    });
  });
}

File: app/src/test/kotlin/com/sslh/sshl/app/TunnelEngineTest.kt
package com.sslh.sshl.app

import com.sslh.sshl.app.model.TunnelConfig
import com.sslh.sshl.app.model.TunnelStatus
import com.sslh.sshl.app.model.TunnelType
import com.sslh.sshl.app.service.TunnelEngine
import org.junit.Assert.*
import org.junit.Test

class TunnelEngineTest {

    @Test
    fun testInitialEngineState() {
        val engine = TunnelEngine()
        assertEquals(TunnelStatus.DISCONNECTED, engine.status)
        assertFalse(engine.isConnected)
        assertFalse(engine.isConnecting)
        assertTrue(engine.logs.isEmpty())
    }

    @Test
    fun testUpdateConfig() {
        val engine = TunnelEngine()
        val newConfig = TunnelConfig(remoteAddr = "10.0.0.1", type = TunnelType.DNS)
        engine.updateConfig(newConfig)

        assertEquals("10.0.0.1", engine.config.remoteAddr)
        assertEquals(TunnelType.DNS, engine.config.type)
    }

    @Test
    fun testAddAndClearLogs() {
        val engine = TunnelEngine()
        engine.addLog("Test log message 1")
        engine.addLog("Test log message 2", isError = true)

        assertEquals(2, engine.logs.size)
        assertEquals("Test log message 1", engine.logs[0].message)
        assertTrue(engine.logs[1].isError)

        engine.clearLogs()
        assertTrue(engine.logs.isEmpty())
    }

    @Test
    fun testStartAndStopTunnel() {
        val engine = TunnelEngine()
        engine.startTunnel()

        // Wait brief moment for thread execution
        Thread.sleep(300)

        assertTrue(engine.logs.size > 0)

        engine.stopTunnel()
        assertEquals(TunnelStatus.DISCONNECTED, engine.status)
    }
}
